### DISCUSSION TITLE
Terrible compression for near-zero float64 columns. Can this be improved?

### DISCUSSION THREAD

**Discussion by alex-immersive** [ClickHouse/ClickHouse#61724] [Q&A] [answered]
Hello Clickhouse community!

During some experimenting I have noticed that float64 compresses terribly, or not at all, when the values are near zero.

Here is an example table:

```
snapshot_time datetime CODEC (Delta, ZSTD),
    key_1 LowCardinality(String),
    key_2 LowCardinality(String),
    key_3 LowCardinality(String),
    key_4 LowCardinality(String),
    value_1         Float64,
    value_2         Float64,
    value_3         Float64 )
    ENGINE = MergeTree()
      ORDER BY (key_4, key_3, key_2, key_1, snapshot_time)
      PRIMARY KEY (key_4, key_3, key_2, key_1, snapshot_time);
```

Here are the compression rates for the table with data loaded:
| name | compressed\_size | uncompressed\_size | ratio |
| :--- | :--- | :--- | :--- |
| snapshot_time | 78.25 KiB | 45.39 MiB | 594.01 |
| key_1 | 424.17 KiB | 90.79 MiB | 219.18 |
| key_2 | 57.30 KiB | 11.37 MiB | 203.24 |
| key_3 | 58.99 KiB | 11.37 MiB | 197.41 |
| key_4 | 60.42 KiB | 11.37 MiB | 192.73 |
| value_1 | 91.18 MiB | 90.79 MiB | 1 |
| value_2 | 91.18 MiB | 90.79 MiB | 1 |
| value_3 | 91.18 MiB | 90.79 MiB | 1 |

The values in this db are highly correlated when sorted by the sorting keys. Here are some example values for one specific set of keys: (time descending)

| value_1 | value_2 | value_3 |
| :--- | :--- | :--- |
| 0.46876019283001774 | -0.019383375420297333 | 0.0019734242713218553 |
| 0.46989613941719954 | -0.016619783315463654 | 0.002170882961624432 |
| 0.465690885973469 | -0.024414204185399302 | 0.001171840555786896 |
| 0.4684201425294021 | -0.019542915560245122 | 0.0017774320893827467 |
| 0.4684560533984955 | -0.02128932044786558 | 0.0019169218567764434 |
| 0.4677124054863484 | -0.022976284295027072 | 0.0018634043614069173 |
| 0.46737444596488803 | -0.023238170774430578 | 0.001574820336587568 |
| 0.4686520688613266 | -0.02198419398397916 | 0.0019613853682474702 |
| 0.4675132108522501 | -0.024368861304646605 | 0.001649146072317087 |
| 0.4691827125747546 | -0.022473022539355514 | 0.002249102205649278 |

I understand this is using default compression, so I did some testing with the following codecs:

```
CODEC(FPC(28), ZSTD(22))
CODEC(FPC(28))
CODEC(Gorilla)
CODEC(Gorilla, ZSTD(22))
CODEC(Delta, ZSTD(22))
CODEC(LZ4HC(12))        
CODEC(ZSTD(22))          
```

The best result I could get was with `CODEC(Delta, ZSTD(22))` which gave a ratio of 1.31. Should I be expecting better than this?  Is there a better codec I could be using for this type of data?

**Comment by den-crane** [comment]
If you don't need such exact accuracy you can round them to some digit or you can use float32.

>when the values are near zero.

I doubt that it's related.

**Reply by alex-immersive** [reply]
Thanks for the reply. Ideally we don't want to lose the precision as this is financial data. Can you suggest any other options instead? Keen on making this better if possible.

**Comment by alex-immersive** [comment, marked as answer]
Update for anyone interested. The best compression I could get was by truncating to 8dp (our min acceptable precision), then multiplying by 10^8 to and storing as int64 with codec (T64, ZSTD). This gets an average ratio of 2.8. I should note that query performance did decrease very slightly, which I expect is due to slower decompression.

**Reply by UnamedRus** [reply]
You can also check rounding in that way

```
SELECT
    CAST('5.79123453132', 'Float64') AS x,
    reinterpretAsFloat64(bitAnd(reinterpretAsUInt64(x), 0xFFFFFFFFFF000000)) AS rounded


┌─────────────x─┬───────────rounded─┐
│ 5.79123453132 │ 5.791234523057938 │
└───────────────┴───────────────────┘
```
